### PR TITLE
GLTFLoader: Multiple Associations & Self Lookup

### PR DESCRIPTION
**Pull request by ghost**
Related PR #21737

Allows for multiple glTF associations per Three.js Object.
Provides glTF self look up indices on a Three.js Object in order to facilitate easy identification of the objects glTF counterparts.

When Three.js loads a glTF file it combines the glTF node, mesh, and primitive components into single objects in some cases, this change provides a map of those combinations in order to preserve an objects relationship to the original glTF file.

### CODE REVIEW

**Comment by mrdoob** [comment]
@takahirox I recall you also used this feature?

**Comment by takahirox** [comment]
Yes, I need multiple associations so I made #21737 PR (it's not an issue but a PR).

To clarify, is the difference from #21737 to have a reference index from Three.js object userData to glTF?

**Comment by timmmeh** [comment]
> Yes, I need multiple associations so I made #21737 PR (it's not an issue but a PR).
> 
> To clarify, is the difference from #21737 to have a reference index from Three.js object userData to glTF?

Yes that ended up being the major difference, I started by ensuring that Three.js does not combing multiple glTF nodes in a Three object, but I don't think we need to go that far, I'm pretty sure just allowing multiple associations is good enough. If you would like to merge this extra bit into your PR I'm sure that will work out just fine for us.

**Comment by mrdoob** [comment]
@takahirox What do you want to do?

**Comment by takahirox** [comment]
If mine looks ok, I'm happy if we can merge mine first.

**Comment by timmmeh** [comment]
Yes please merge @takahirox changes, I have started pre-merging my changes with those and will update the PR shortly

**Comment by timmmeh** [comment]
Refactors 'multiple-association' into a helper method `_addAssociation()`

Fixes association issue where a Three `Mesh` object that represents a glTF Node, Mesh, and Primitive is missing the `nodes` mapping. Resolved by propagating the association mappings when an object is cloned.

Adds self-lookup to userData:
```
ThreeObject {
  ...
  userData: {
    associations: {
      nodes: nodesIndex,
      meshes: meshedIndex,
      primitives: primitivesIndex,
    }
  }
}
```

File: examples/jsm/loaders/GLTFLoader.js [hunk 1]
```diff
@@ -2411,6 +2422,31 @@ class GLTFParser {
...
+	_addAssociation( object, type, index ) {
+
+		const mappings = this.associations.get( object ) || {};
+		mappings[ type ] = index;
+		this.associations.set( object, mappings );
+
+		// Gives the object the ability to find itself in the glTF data structure
+		// of the given type, in other words this indicates which glTF object
+		// the Three object was generated from and/or represents.
+		object.userData = object.userData || {};
```
Comment (takahirox) [inline examples/jsm/loaders/GLTFLoader.js:2444]: What type object doesn't have `userData` by default?

**Comment by takahirox** [comment]
Yeah this change allows cloned objects to look up the associated glTF data. This change makes me wonder

1. if we still need `associations` in the parser if we provide a way to self-lookup?
2. if there are any historical reasons why we didn't store the mappings in `object.userData` but we provided `parser.associations` instead? @donmccurdy do you remember?

**Comment by donmccurdy** [comment]
> if there are any historical reasons why we didn't store the mappings in object.userData but we provided parser.associations instead?

I do worry about putting things into `.userData` that aren't JSON serializable. It may cause issues for the Editor, or for users exporting the model later — not all exports will go through GLTFExporter. By comparison the parser's `associations` object won't interfere with any of that.

**Comment by timmmeh** [comment]
Ok no problem, the only thing of substance left here is the fix I made on top of @takahirox 's change that makes sure the `node` mapping makes it into the associations, this was getting left out in the case where a Node, mesh, and single-primitive get combined into one Three.js object. I will break that out into a new PR.

**Comment by timmmeh** [comment]
Break out PR added here #22583

**Pull request closed without merging**